sofar/sofar.py

```diff
@@ -0,0 +1,2125 @@
+
+    # convert arrays to scalars if they do not store data that is usually used
+    # as scalar metadata, e.g., the SamplingRate
+    data_keys = ["Data_IR", "Data_Real", "Data_Imag", "Data_SOS" "Data_Delay"]
+    if value.size == 1 and key not in data_keys:
+        value = value[0]
+
+    return value
+
+
+def _add_custom_api_entry(sofa, key, value, flags, dimensions, dtype):
+    """
+    Add custom entry to the sofa._convention and permanently save it in
+    sofa._custom
+
+    Parameters
+    ----------
+    sofa : Sofa
+    key : str
+        name of the entry
+    flags, dimensions, dtype : any
+        as in sofa._convention
+    """
+    # create custom API if it not exists
+    sofa._protected = False
+    if not hasattr(sofa, "_custom"):
+        sofa._custom = {}
+
+    # lower case letters to indicate custom dimensions
+    if dimensions is not None:
+        dimensions = [d.upper() if d.upper() in "ERMNCIS" else d.lower()
+                      for d in dimensions]
+        dimensions = "".join(dimensions)
+
+    # add user entry to custom API
+    sofa._custom[key] = {
+        "flags": flags,
+        "dimensions": dimensions,
+        "type": dtype,
+        "default": None,
+        "comment": ""}
+    sofa._update_convention(version="match")
+
+    # add attribute to object
+    setattr(sofa, key, value)
+    sofa._protected = True
+
+
+def _verify_convention_and_version(version, version_in, convention):
+    """
+    Verify if convention and version exist and return version
+
+    Parameters
+    ----------
+    version : str
+        'latest', 'match', version string (e.g., '1.0')
+    version_in : str
+        The version to be checked against
+    convention : str
+        The name of the convention to be checked
+
+    Returns
+    -------
+    version_out : str
+        The version to be used depending on `version`, and `version_in`
+    """
+
+    # check if the convention exists in sofar
+    if convention not in sf.list_conventions(False, "name"):
+        raise ValueError(
+            f"Convention {convention} does not exist")
+
+    name_version = sf.list_conventions(False, "name_version")
+
+    if version == "latest":
+        # get latest version (comes last)
+        for versions in name_version:
+            if versions[0] == convention:
+                version_out = versions[1]
+    else:
+        # check which version is wanted
+        if version == "match":
+            match = version_in
+        else:
+            match = version
+
+        version_out = None
+        for versions in name_version:
+            # check if convention and version match
+            if versions[0] == convention \
+                    and str(float(versions[1])) == match:
+                version_out = str(float(versions[1]))
+
+        if version_out is None:
+            raise ValueError(
+                f"Version {match} does not exist. Try version='latest'")
+
+    return version_out
+
+
+def _nd_array(array, ndim):
```

Comment:
maybe rename to `_atleast_nd` analogous to numpy, since no new empty array is created here, but an existing one is extended to a specific dimension.
```suggestion
def _atleast_nd(array, ndim):
```

Reply:
yes, changed...